presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/CachingJdbcClient.java

```diff
@@ -0,0 +1,289 @@
+    }
+
+    public CachingJdbcClient(JdbcClient delegate, Duration metadataCachingTtl, boolean cacheMissing)
+    {
+        this.delegate = requireNonNull(delegate, "delegate is null");
+        CacheBuilder<Object, Object> cacheBuilder = CacheBuilder.newBuilder()
+                .expireAfterWrite(metadataCachingTtl.toMillis(), TimeUnit.MILLISECONDS);
+        columnsCache = cacheBuilder.build();
+        tableHandleCache = cacheBuilder.build();
+        schemaNamesCache = cacheBuilder.build();
+        tableNamesCache = cacheBuilder.build();
+        this.cacheMissing = cacheMissing;
+    }
+
+    @Override
+    protected JdbcClient getDelegate()
+    {
+        return delegate;
+    }
+
+    @Override
+    public Set<String> getSchemaNames(JdbcIdentity identity)
+    {
+        Set<String> schemas = schemaNamesCache.getIfPresent(identity);
+        if (schemas != null) {
+            return schemas;
+        }
+
+        schemas = getDelegate().getSchemaNames(identity);
+        schemaNamesCache.put(identity, schemas);
+        return schemas;
+    }
+
+    @Override
+    public List<SchemaTableName> getTableNames(JdbcIdentity identity, Optional<String> schema)
+    {
+        TableNamesCacheKey key = new TableNamesCacheKey(identity, schema);
+        List<SchemaTableName> tableNames = tableNamesCache.getIfPresent(key);
+        if (tableNames != null) {
+            return tableNames;
+        }
+        tableNames = getDelegate().getTableNames(identity, schema);
+        tableNamesCache.put(key, tableNames);
+        return tableNames;
+    }
+
+    @Override
+    public List<JdbcColumnHandle> getColumns(ConnectorSession session, JdbcTableHandle tableHandle)
+    {
+        ColumnsCacheKey key = new ColumnsCacheKey(JdbcIdentity.from(session), tableHandle);
+        List<JdbcColumnHandle> columns = columnsCache.getIfPresent(key);
+        if (columns != null) {
+            return columns;
+        }
+        columns = getDelegate().getColumns(session, tableHandle);
+        columnsCache.put(key, columns);
+        return columns;
+    }
+
+    @Override
+    public Optional<JdbcTableHandle> getTableHandle(JdbcIdentity identity, SchemaTableName schemaTableName)
+    {
+        TableHandleCacheKey key = new TableHandleCacheKey(identity, schemaTableName);
+        Optional<JdbcTableHandle> cachedTableHandle = tableHandleCache.getIfPresent(key);
+        if (cachedTableHandle != null) {
+            return cachedTableHandle;
+        }
+        Optional<JdbcTableHandle> tableHandle = getDelegate().getTableHandle(identity, schemaTableName);
+        if (!tableHandle.isPresent() || cacheMissing) {
+            tableHandleCache.put(key, tableHandle);
+        }
+        return tableHandle;
+    }
+
+    @Override
+    public void commitCreateTable(JdbcIdentity identity, JdbcOutputTableHandle handle)
+    {
+        getDelegate().commitCreateTable(identity, handle);
+        invalidateTablesCache();
+    }
+
```

Comment:
`finishInsertTable` should invalidate table handle cache, in case the table handle stores any auxiliary info